core/src/main/java/com/scalar/db/api/ScanBuilder.java

```diff
@@ -205,9 +218,253 @@ public Scan build() {
+      checkNotNull(condition);
+      where.and(condition);
+      return new BuildableScanWithOngoingWhereAnd(this);
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereAnd and(OrConditionSet orConditionSet) {
+      checkNotNull(orConditionSet);
+      where.and(orConditionSet);
+      return new BuildableScanWithOngoingWhereAnd(this);
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereOr or(ConditionalExpression condition) {
+      checkNotNull(condition);
+      where.or(condition);
+      return new BuildableScanWithOngoingWhereOr(this);
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereOr or(AndConditionSet andConditionSet) {
+      checkNotNull(andConditionSet);
+      where.or(andConditionSet);
+      return new BuildableScanWithOngoingWhereOr(this);
+    }
+  }
+
+  public static class BuildableScanWithOngoingWhereAnd extends BuildableScanWithWhere
+      implements And<BuildableScanWithOngoingWhereAnd> {
+
+    private BuildableScanWithOngoingWhereAnd(BuildableScanWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanWithOngoingWhereAnd(
+        BuildableScan buildable, OrConditionSet orConditionSet) {
+      super(buildable);
+      where.and(orConditionSet);
+    }
+
+    private BuildableScanWithOngoingWhereAnd(
+        BuildableScan buildable, Set<OrConditionSet> orConditionSets) {
+      super(buildable);
+      where.and(orConditionSets);
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereAnd and(ConditionalExpression condition) {
+      checkNotNull(condition);
+      where.and(condition);
+      return this;
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereAnd and(OrConditionSet orConditionSet) {
+      checkNotNull(orConditionSet);
+      where.and(orConditionSet);
+      return this;
+    }
+  }
+
+  public static class BuildableScanWithOngoingWhereOr extends BuildableScanWithWhere
+      implements Or<BuildableScanWithOngoingWhereOr> {
+
+    private BuildableScanWithOngoingWhereOr(BuildableScanWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanWithOngoingWhereOr(
+        BuildableScan buildable, AndConditionSet andConditionSet) {
+      super(buildable);
+      where.or(andConditionSet);
+    }
+
+    private BuildableScanWithOngoingWhereOr(
+        BuildableScan buildable, Set<AndConditionSet> andConditionSets) {
+      super(buildable);
+      where.or(andConditionSets);
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereOr or(ConditionalExpression condition) {
+      checkNotNull(condition);
+      where.or(condition);
+      return this;
+    }
+
+    @Override
+    public BuildableScanWithOngoingWhereOr or(AndConditionSet andConditionSet) {
+      checkNotNull(andConditionSet);
+      where.or(andConditionSet);
+      return this;
+    }
+  }
+
+  public static class BuildableScanWithWhere extends BuildableScan {
```

Comment:
No, it cannot since I think we should be able to do something like this.

```java
    BuildableScanWithWhere buildable =
        Scan.newBuilder()
            .namespace("cassandrans")
            .table("lineitem")
            .partitionKey(Key.ofInt("l_orderkey", 807))
            .where(ConditionBuilder.column("l_discount").isGreaterThanDouble(0.04));

    Scan scan1 = buildable.build();
    Scan scan2 = buildable.start(Key.ofInt("l_linenumber", 2)).build();
```

JFYI: I noticed `BuildableScanWithWhere` should not inherit `BuildableScan` above since `BuildableScan` does not build a scan with the where clause as long as users intentionally use (or cast) the `WithWhere` object, and users hardly know they should use it. I will fix it. Thank you for the chance to notice it.

Reply:
Sorry, I misunderstood it. We can make it `abstract`, but still I think it can be as is because we do not intend that users always willingly use the base one, but rather intend to properly use both the base one and inherited ongoing ones based on the situation.